internal/inspect/inspect.go

```diff
@@ -0,0 +1,156 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+package inspect
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+
+	"github.com/cartesi/rollups-node/internal/node/advancer/machines"
+	. "github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/cartesi/rollups-node/internal/nodemachine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+var (
+	ErrInvalidMachines = errors.New("machines must not be nil")
+	ErrNoApp           = errors.New("no machine for application")
+)
+
+type Inspector struct {
+	machines Machines
+}
+
+type InspectorResponse struct {
+	Status     string   `json:"status"`
+	Exception  string   `json:"exception"`
+	Reports    []string `json:"reports"`
+	InputIndex uint64   `json:"processed_input_count"`
+}
+
+// New instantiates a new Inspect.
+func New(machines Machines) (*Inspector, error) {
+	if machines == nil {
+		return nil, ErrInvalidMachines
+	}
+
+	return &Inspector{machines: machines}, nil
+}
+
+func (inspect *Inspector) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var (
+		dapp    Address
+		payload []byte
+		err     error
+		reports []string
+		status  string
+	)
+
+	if r.PathValue("dapp") == "" {
+		slog.Info("Bad request",
+			"service", "inspect",
+			"err", "Missing application address")
+		http.Error(w, "Missing application address", http.StatusBadRequest)
+		return
+	}
+
+	dapp = common.HexToAddress(r.PathValue("dapp"))
+	if r.Method == "POST" {
+		payload, err = io.ReadAll(r.Body)
+		if err != nil {
+			slog.Info("Bad request",
+				"service", "inspect",
+				"err", err)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	} else {
+		if r.PathValue("payload") == "" {
+			slog.Info("Bad request",
+				"service", "inspect",
+				"err", "Missing payload")
+			http.Error(w, "Missing payload", http.StatusBadRequest)
+			return
+		}
+		payload, err = hexutil.Decode(r.PathValue("payload"))
+		if err != nil {
+			slog.Info("Internal server error",
+				"service", "inspect",
+				"err", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
A failure in decoding would suggest a wrongly encoded payload, right? Something like this:
```suggestion
			http.Error(w, err.Error(), http.StatusBadRequest)
```